### ISSUE TITLE
build fails because of null conversion error

### ISSUE THREAD

**Issue by anna-prins1** [aws/aws-sdk-cpp#3033]
### Describe the bug

When building the sdk using the steps in [this tutorial](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/setup-linux.html), the build fails on the step `Building CXX object src/aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/ub_core.cpp.o`. 

### Expected Behavior

Expected behavior would be to make without error.

### Current Behavior

Here's the trace:
```
In file included from /home/annaprins/sdk_build/src/aws-cpp-sdk-core/ub_core.cpp:26:
/home/annaprins/aws-sdk-cpp/src/aws-cpp-sdk-core/source/client/RequestCompression.cpp: In member function ‘iostream_outcome Aws::Client::RequestCompression::compress(std::shared_ptr<std::basic_iostream<char> >, const Aws::Client::CompressionAlgorithm&) const’:
/home/annaprins/aws-sdk-cpp/src/aws-cpp-sdk-core/source/client/RequestCompression.cpp:102:19: error: converting to non-pointer type ‘int’ from NULL [-Werror=conversion-null]
  102 |         int ret = Z_NULL;
      |                   ^~~~~~
/home/annaprins/aws-sdk-cpp/src/aws-cpp-sdk-core/source/client/RequestCompression.cpp: In member function ‘Aws::Utils::Outcome<std::shared_ptr<std::basic_iostream<char> >, bool> Aws::Client::RequestCompression::uncompress(std::shared_ptr<std::basic_iostream<char> >, const Aws::Client::CompressionAlgorithm&) const’:
/home/annaprins/aws-sdk-cpp/src/aws-cpp-sdk-core/source/client/RequestCompression.cpp:221:19: error: converting to non-pointer type ‘int’ from NULL [-Werror=conversion-null]
  221 |         int ret = Z_NULL;
      |                   ^~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [src/aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/build.make:977: src/aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/ub_core.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:987: src/aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

### Reproduction Steps

I used this line to cmake: 
`cmake ../aws-sdk-cpp -DCMAKE_BUILD_TYPE=Debug -DCMAKE_PREFIX_PATH=/usr/local/ -DCMAKE_INSTALL_PREFIX=/usr/local/ -DBUILD_ONLY="logs"`
and then it failed on `make`
I tried both Debug and Release build types. I don't know why I thought that would change anything, but I thought it was worth a shot.. It fails the same way with both.

### AWS CPP SDK version used

latest

### Compiler and Version used

gcc 14.1.1

### Operating System and version

Fedora 40

**Comment by jmklix** [comment]
I can't seem to reproduce this error that you're seeing. I made a dockerfile to see if it's reproducible on Fedora 40 and I didn't run into the build error that you're seeing. Can you take a look at the dockerfile below and see if you're able to get the sdk build in your environment or modify it to reproduce the same error that you get?

### Dockerfile

layer.tar was downloaded from [here](https://github.com/fedora-cloud/docker-brew-fedora/tree/e1c3015854b7ccfd458331f5a7a77886c26d7014/x86_64)

```Dockerfile
# This is a Jinja2 compatible Dockerfile template
FROM scratch
LABEL maintainer="Clement Verna <cverna@fedoraproject.org>"
ENV DISTTAG=f40container FGC=f40 FBR=f40
ADD layer.tar /
CMD ["/bin/bash"]

RUN yum update -y
RUN yum install -y git cmake gcc-c++ libcurl-devel zlib-devel openssl-devel

RUN git clone --depth 1 --recurse-submodules https://github.com/aws/aws-sdk-cpp && \
    cd aws-sdk-cpp && \
    mkdir build && \
    cd build

```

### Commands

```
docker build -t 3033 .
docker run --name 3033_cont -it 3033 /bin/bash
```
Then inside the docker container I ran these commands
```
cd aws-sdk-cpp/build
cmake .. -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/aws-sdk-cpp/install/ -DBUILD_ONLY="logs"
make
```

**Comment by anna-prins1** [comment]
I was able to build it using the Dockerfile you sent me on in a container on my Mac. But on my Fedora machine on the bare OS I am not able to get it to build. I tried to use the same commands you used in the Dockerfile and the ones you used to build, but I still got the same error. So I diff-ed the output of the cmake command for each, and it turns out my Fedora machine had zlib version 1.3.0 on it and the docker container had v 1.3.1. I updated zlib and built successfully. So, something must have been broken in the last version of zlib.
